internal/server/evaluation/evaluator.go

```diff
@@ -131,81 +124,13 @@ func (e *Evaluator) Evaluate(ctx context.Context, r *flipt.EvaluationRequest) (r
 
 		lastRank = rule.Rank
 
-		constraintMatches := 0
-
-		// constraint loop
-		for _, c := range rule.Constraints {
-			v := r.Context[c.Property]
-
-			var (
-				match bool
-				err   error
-			)
-
-			switch c.Type {
-			case flipt.ComparisonType_STRING_COMPARISON_TYPE:
-				match = matchesString(c, v)
-			case flipt.ComparisonType_NUMBER_COMPARISON_TYPE:
-				match, err = matchesNumber(c, v)
-			case flipt.ComparisonType_BOOLEAN_COMPARISON_TYPE:
-				match, err = matchesBool(c, v)
-			case flipt.ComparisonType_DATETIME_COMPARISON_TYPE:
-				match, err = matchesDateTime(c, v)
-			default:
-				resp.Reason = flipt.EvaluationReason_ERROR_EVALUATION_REASON
-				return resp, errs.ErrInvalid("unknown constraint type")
-			}
-
-			if err != nil {
-				resp.Reason = flipt.EvaluationReason_ERROR_EVALUATION_REASON
-				return resp, err
-			}
-
-			if match {
-				e.logger.Debug("constraint matches", zap.Reflect("constraint", c))
-
-				// increase the matchCount
-				constraintMatches++
-
-				switch rule.SegmentMatchType {
-				case flipt.MatchType_ANY_MATCH_TYPE:
-					// can short circuit here since we had at least one match
-					break
-				default:
-					// keep looping as we need to match all constraints
-					continue
-				}
-			} else {
-				// no match
-				e.logger.Debug("constraint does not match", zap.Reflect("constraint", c))
-
-				switch rule.SegmentMatchType {
-				case flipt.MatchType_ALL_MATCH_TYPE:
-					// we can short circuit because we must match all constraints
-					break
-				default:
-					// keep looping to see if we match the next constraint
-					continue
-				}
-			}
-		} // end constraint loop
-
-		switch rule.SegmentMatchType {
-		case flipt.MatchType_ALL_MATCH_TYPE:
-			if len(rule.Constraints) != constraintMatches {
-				// all constraints did not match, continue to next rule
-				e.logger.Debug("did not match ALL constraints")
-				continue
-			}
+		matched, err := doConstraintsMatch(e.logger, r.Context, rule.Constraints, rule.SegmentMatchType)
```

Comment:
```suggestion
		matched, err := doConstraintsMatch(e.logger, r.Context, rule.Constraints, rule.SegmentMatchType)
```

I think we could simplify this method call as well: `e.matchConstraints(r.Context, rule)` since you can then get the `Constraints` and `SegmentMatchType` off of the rule

Reply:
Yeah I am sharing this common function with the `Boolean` evaluation logic which doesn't have a rule. I suppose I can use an interface that exposes getting `Constraints`, and `SegmentMatchType`?